### ISSUE TITLE
We use too much .fold(), there is libimagutil::iter::FoldResult!

### ISSUE THREAD

**Issue by matthiasbeyer** [matthiasbeyer/imag#777]
The documentation has to be improved for `FoldResult`: https://matthiasbeyer.github.io/imag/libimagutil/iter/trait.FoldResult.html#tymethod.fold_result

But then we should _really_ use it:

``` bash
$ ag "fold\(" -l
libimagentrylink/src/internal.rs
libimagentrylist/src/listers/table.rs
libimagentrylist/src/listers/core.rs
bin/src/main.rs
libimagutil/src/iter.rs
libimaginteraction/src/ui.rs
libimagref/src/lister.rs
```

I assume almost _all_ cases can be refactored with `FoldResult`.

---

This is a beginner-task, but not if you're a complete rust beginner.

**Comment by matthiasbeyer** [comment]
Note to self: Submitted to this-week-in-rust as easy issue that might be interesting for contributors.

**Comment by matthew-piziak** [comment]
@matthiasbeyer Thanks for identifying easy issues and posting them to this-week-in-rust. I'll take on this refactor. Next update within 48h.

**Comment by matthew-piziak** [comment]
@matthiasbeyer I have to admit I'm quite confused by the behavior of the `FoldResult` trait. It throws away the accumulator while `fold`ing, which makes it a degenerate fold, with the behavior (in Haskell) of `last . mapM`. In contrast, [`itertools::fold_results`](https://docs.rs/itertools/*/itertools/trait.Itertools.html#method.fold_results) is a true fold over `Result`s. In plain English, `FoldResult` returns the first error, unless there are none in which case the last success, unless there are none in which case the default.

I do not believe it is the case that all uses of `fold` in the codebase have vacuous accumulators that can be refactored with `FoldResult`.

**Comment by matthew-piziak** [comment]
I wrote the following test cases while trying to understand `FoldResult`:

```rust
#[test]
fn test_fold_result_success() {
    let v = vec![1, 2, 3];
    let result: Result<usize, &str> = v.iter().fold_result(5, |item| Ok(*item));
    assert_eq!(result, Ok(3));
}

#[test]
fn test_fold_result_failure() {
    let v = vec![1, 2, 3];
    let result: Result<usize, &str> = v.iter().fold_result(5, |_| Err("failure"));
    assert_eq!(result, Err("failure"));
}

#[test]
fn test_fold_result_default() {
    let v: Vec<usize> = vec![];
    let result: Result<usize, &str> = v.iter().fold_result(5, |_| Err("failure"));
    assert_eq!(result, Ok(5));
}

#[test]
fn test_fold_defresult_success() {
    let v = vec![1, 2, 3];
    let result: Result<usize, &str> = v.iter().fold_defresult(|item| Ok(*item));
    assert_eq!(result, Ok(3));
}

#[test]
fn test_fold_defresult_failure() {
    let v = vec![1, 2, 3];
    let result: Result<usize, &str> = v.iter().fold_defresult(|_| Err("failure"));
    assert_eq!(result, Err("failure"));
}

#[test]
fn test_fold_defresult_default() {
    let v: Vec<usize> = vec![];
    let result: Result<usize, &str> = v.iter().fold_defresult(|_| Err("failure"));
    assert_eq!(result, Ok(0));
}
```

**Comment by matthiasbeyer** [comment]
> @matthiasbeyer I have to admit I'm quite confused by the behavior of the FoldResult trait. It throws away the accumulator while folding, which makes it a degenerate fold, with the behavior (in Haskell) of last . mapM. 

Yes, that's exactly what we want to do with the `FoldResult` type. We use it when we have to work on a list of things, but do not care if it succeeds (besides from going on) but stop when the first error occurs. "Throwing away" the accumulator is the right thing to do in this case. The `FoldResult` functionality is used in places where we return things like `Result<(), ErrorType>`.

> In contrast, itertools::fold_results is a true fold over Results. 

The `itertools::fold_results()` is rather orthogonal to our `FoldResult` (also notice fold_result**s** vs. FoldResul**t**). From what I see: The `itertool::fold_results()` folds over (e.g.) a `Vec<Result<_, _>>`. We do fold over a `Vec<_>`, but with a function that returns a `Result<_, _>` - that's the difference!

> I do not believe it is the case that all uses of fold in the codebase have vacuous accumulators that can be refactored with FoldResult.

That statement is true. But I think there are _some_ which can be refactored!

**Comment by matthiasbeyer** [comment]
Ah, feel free to submit these test cases as tests in the appropriate file, if you want to! :+1:

**Comment by matthew-piziak** [comment]
@matthiasbeyer Will do! In fact, I may add them as examples. Documentation tests will still validate them, but they'll be more visible to users.

**Comment by matthew-piziak** [comment]
On second thought I'm leaving them as unit tests for now. I had some weird import issues in documentation tests. I can always move them over later.

**Comment by matthew-piziak** [comment]
>We use it when we have to work on a list of things, but do not care if it succeeds (besides from going on)

Oh! I was assuming that the last success was always important.

**Comment by matthiasbeyer** [comment]
> Oh! I was assuming that the last success was always important.

Yes and No. The last success is only important as we know that everything succeeded, but not as a value itself. That's why the accumulator is thrown away in every iteration: the value does not matter, just the information that it succeeded.

**Comment by matthew-piziak** [comment]
Here's the first refactor target, in `internal.rs`:

```rust
let new_links = try!(links_into_values(new_links)
                     .into_iter() 
                     .fold(Ok(vec![]), |acc, elem| {
                        acc.and_then(move |mut v| {
                            elem.map_err_into(LEK::InternalConversionError)
                                .map(|e| {
                                    v.push(e);
                                    v
                                })
                        })
                    }));
```

The `Default` of `Vec<_>` is `vec![]`, so let's go ahead and use `fold_defresult`.

```rust
let new_links = try!(links_into_values(new_links)
                         .into_iter()
                         .fold_defresult(move |mut v| {
                             elem.map_err_into(LEK::InternalConversionError)
                                 .map(|e| {
                                     v.push(e);
                                     v
                                 })
                         }));
```

Here the naiive refactoring doesn't work because the inner function still uses a reference to `elem`. I'll have to think about this some more.

**Comment by matthew-piziak** [comment]
@matthiasbeyer Based on your description, wouldn't this implementation be simpler?

```rust
pub trait FoldResult: Sized {
    type Item;

    /// Processes all contained items, returning the first error if it exists.
    fn fold_result<R, E, F>(self, mut func: F) -> Result<(), E>
        where F: FnMut(Self::Item) -> Result<R, E>;
}

impl<X, I: Iterator<Item = X>> FoldResult for I {
    type Item = X;

    fn fold_result<R, E, F>(self, mut func: F) -> Result<(), E>
        where F: FnMut(Self::Item) -> Result<R, E>
    {
        for item in self {
            try!(func(item));
        };
        Ok(())
    }
}
```

**Comment by matthew-piziak** [comment]
After all, then you can perform these operations:

```rust
#[test]
fn test_fold_result_success() {
    let v = vec![1, 2, 3];
    let mut accum = vec![];
    let result: Result<(), &str> = v.iter().fold_result(|item| {
        accum.push(*item * 2);
        Ok(*item)
    });
    assert_eq!(result, Ok(()));
    assert_eq!(accum, vec![2, 4, 6]);
}

#[test]
fn test_fold_result_failure() {
    let v: Vec<usize> = vec![1, 2, 3];
    let mut accum: Vec<usize> = vec![];
    let result: Result<(), &str> = v.iter().fold_result(|item| if *item == 2 {
        Err("failure")
    } else {
        accum.push(*item * 2);
        Ok(*item)
    });
    assert_eq!(result, Err("failure"));
}
```

**Comment by matthiasbeyer** [comment]
@matthew-piziak why are you submitting these code snippets here as comments and not as PR? :smile:

**Comment by matthew-piziak** [comment]
@matthiasbeyer Hah, fair enough! Here's one.
